docs/proposals/008-tuples.md

```diff
@@ -0,0 +1,137 @@
+SP #008 - Tuples
+==============
+
+Now that we have variadic generics in the language following [SP #007], we should now be able to support `Tuple` type as a core language feature.
+`Tuple` types are useful in many places to reduce boilerplate in user code, such as in function return types to eliminate the need of defining
+`struct`s that are used only for invoking the function. Adding `Tuple` types to Slang will also simplify interop with other languages such as Python
+and C++ that have tuple types.
+
+Status
+------
+
+Author: Yong He
+
+Implementation: In-progress.
+
+Reviewed by: N/A
+
+Background
+----------
+
+Tuple type is widely supported in almost all of the modern programming languages including C++, C#, Swift, Rust, Python. Supporting tuple types
+in Slang will bring the language to parity with other languages and allow users to practice the same coding idioms in Slang, and allow Slang code
+to interop more directly with other parts of the user application written in other languages.
+
+
+Proposed Approach
+-----------------
+
+With variadic generics support, we can now easily define a Tuple type in stdlib as:
+```
+__generic<each T>
+__magic_type(TupleType)
+struct Tuple
+{
+    __intrinsic_op($(0))
+    __init(expand each T);
+}
+```
+
+This will allow users to instantiate tuple types from their code with `Tuple<T0, T1, T2>(v0, v1, v2)`.
+
+### Constructing Tuple Values
+
+To make it easy to construct tuples, we will define a `makeTuple` function in stdlib as:
+```
+__intrinsic_op($(kIROp_MakeTuple))
+Tuple<expand each T> makeTuple(expand each T values);
+```
+
+With generic argument inferencing, this will enable user to write:
+```
+makeTuple(1, 2.0f) // returns Tuple<int, float>(1, 2.0f)
+```
+
+### Accessing Tuple Elements
```

Comment:
How can we get the size of the Tuple?
Maybe something like t.getCount()?

Reply:
We should be able to do `countof(tupleValue)` just like `countof(T)`.